# Insightful-Q-A-Chatbot

### Google Palm LLM in LangChain

In [1]:
import os
from dotenv import load_dotenv
load_dotenv()

True

In [3]:
from langchain.llms import GooglePalm

api_key = 'GOOGLE_API_KEY' #get this api key from https://makersuite.google.com/

llm = GooglePalm(GOOGLE_API_KEY=os.environ["GOOGLE_API_KEY"], temperature=0.6)

c:\Sahil\portfolio\Insightful-Q-A-Chatbot\venv\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [9]:
# Write a python code for calculator application
code = llm("Write a python code for calculator")
print(code)

```python
# A simple calculator program in Python

# Define the functions for addition, subtraction, multiplication, and division.
def add(a, b):
    return a + b

def subtract(a, b):
    return a - b

def multiply(a, b):
    return a * b

def divide(a, b):
    return a / b

# Get the user's input for the two numbers and the operation to perform.
first_number = int(input("Enter the first number: "))
second_number = int(input("Enter the second number: "))
operation = input("Enter the operation to perform (+, -, *, /): ")

# Perform the operation and print the result.
if operation == "+":
    result = add(first_number, second_number)
elif operation == "-":
    result = subtract(first_number, second_number)
elif operation == "*":
    result = multiply(first_number, second_number)
elif operation == "/":
    result = divide(first_number, second_number)
else:
    print("Invalid operation.")

print("The result is", result)
```


In [6]:
theory = llm("Explain Google Palm model")
print(theory)

The Palm model is a hierarchical recurrent neural network (RNN) designed to process sequential data. It is composed of a stack of RNN layers, each of which is connected to the previous layer via a gating mechanism. This allows the model to learn long-term dependencies in the data.

The Palm model was developed by researchers at Google AI in 2017. It was first used to win the first-place prize in the Google Loungin' competition, a challenge to build a model that could generate text from a prompt. The model has since been used for a variety of tasks, including machine translation, question answering, and text summarization.

The Palm model is a powerful tool for processing sequential data. It is able to learn long-term dependencies in the data, which makes it well-suited for tasks such as machine translation and question answering. However, the model can also be computationally expensive to train, and it may not be suitable for all tasks.

Here is an example of how the Palm model can be 

In [5]:
movies = llm("Write 6 best super hero movies")
print(movies)

1. The Dark Knight (2008)
2. The Avengers (2012)
3. Spider-Man: Into the Spider-Verse (2018)
4. Black Panther (2018)
5. Wonder Woman (2017)
6. The Incredibles (2004)


In [4]:
from langchain.chains import RetrievalQA
from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

### Now let's load data from csv file

In [3]:
from langchain.document_loaders.csv_loader import CSVLoader

# Specify the encoding as 'latin1' to handle non-UTF-8 encoded characters
loader = CSVLoader(file_path='random_questions.csv', source_column="prompt", encoding='latin1')

# Store the loaded data in the 'data' variable
data = loader.load()

# Print the data to verify it loaded correctly
print(data[1])


page_content='prompt: Why should I trust Codebasics?\nresponse: Till now 9000 + learners have benefitted from the quality of our courses. You can check the review section and also we have attached their LinkedIn profiles so that you can connect with them and ask directly.' metadata={'source': 'Why should I trust Codebasics?', 'row': 1}


### Hugging Face Embeddings

https://python.langchain.com/v0.1/docs/integrations/text_embedding/

In [9]:
try:
    from langchain.embeddings import HuggingFaceInstructEmbeddings
    from InstructorEmbedding import INSTRUCTOR
    from sentence_transformers import SentenceTransformer

    # Initialize instructor embeddings using the Hugging Face model
    instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")

    # Embed a query
    el = instructor_embeddings.embed_query("What is your refund policy?")
    print(el)

except ImportError as e:
    print("ImportError: ", e)
    print("Please make sure all dependencies are installed and compatible.")
except Exception as e:
    print("An error occurred: ", e)


load INSTRUCTOR_Transformer
max_seq_length  512
[-0.04389806464314461, 0.0076855444349348545, -0.009231903590261936, 0.02449624612927437, 0.033592283725738525, 0.06253515928983688, 0.010531390085816383, 0.02161242440342903, -0.022162262350320816, 0.02974751405417919, 0.044324278831481934, -0.014968819916248322, 0.04655832052230835, 0.03853655233979225, -0.06129590794444084, -0.04503975063562393, -0.07021934539079666, -0.014154862612485886, -0.02652382105588913, 0.012313466519117355, 0.06716737151145935, -0.005974942352622747, -0.011286315508186817, 0.01893932744860649, 0.011767163872718811, 0.01971540041267872, -0.024538308382034302, 0.027542205527424812, 0.064799465239048, -0.05754318833351135, 0.01590231992304325, -0.03111973963677883, -0.057875435799360275, -0.04131997749209404, -0.024895012378692627, 0.026121458038687706, 0.014952129684388638, 0.00224844878539443, 0.038221005350351334, 0.013710809871554375, -0.004282935988157988, 0.02549811825156212, 0.04529273509979248, -0.0422962

In [10]:
from langchain.embeddings import HuggingFaceInstructEmbeddings

# Initialize instructor embeddings using the Hugging Face model
instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")

e = instructor_embeddings.embed_query("What is your refund policy?")

load INSTRUCTOR_Transformer
max_seq_length  512


In [11]:
len(e)

768

In [12]:
e[:5]

[-0.04389806464314461,
 0.0076855444349348545,
 -0.009231903590261936,
 0.02449624612927437,
 0.033592283725738525]

As we can see above, a list of size 768 is embedded for the sentence "What is your return policy." It is assumed that the numbers in this list represent the meaning of "What is your refund policy" rather than providing an intuitive grasp of what it is.

### Vector store using FAISS

In [13]:
from langchain.vectorstores import FAISS

# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=data,
                                 embedding=instructor_embeddings)

# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)

In [ ]:
rdocs = retriever.get_relevant_documents("how about job placement support?")
rdocs

[Document(page_content='prompt: Do you provide any job assistance?\nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.', metadata={'source': 'Do you provide any job assistance?', 'row': 11}),
 Document(page_content='prompt: Will this course guarantee me a job?\nresponse: We created a much lighter version of this course on YouTube available for free (click this link) and many people gave us feedback that they were able to fetch jobs (see testimonials). Now this paid course is at least 5x better than the YouTube course which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.', metadata={'source': 'Will this 

: 

The retriever that was developed using FAISS and hugging face embedding can now extract relevant documents from our initial CSV file knowledge store.

##### FAISS and embracing face embeddings were more suitable

In [14]:
# google_palm_embeddings = GooglePalmEmbeddings(google_api_key=api_key)

# from langchain.vectorstores import Chroma
# vectordb = Chroma.from_documents(data,
#                            embedding=google_palm_embeddings,
#                            persist_directory='./chromadb')
# vectordb.persist()

### Create RetrievalQA chain

In [ ]:
from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            # chain_type_kwargs=chain_type_kwargs)

: 

In [ ]:
response = chain('Do you provide job assistance and also do you provide job gurantee?')
response

: 

In [ ]:
response = chain('Do you have any plans of launching food course ?')
response

: 

### Create RetrievalQA chain along with prompt template

In [ ]:
from langchain.prompts import PromptTemplate

prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}

: 

In [ ]:
from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs={"prompt": PROMPT})

: 

In [ ]:
response = chain('Do you have any plans of launching food course ?')
response

: 

**As you can see above, the answer of question comes from two different FAQs within our csv file and it is able to pull those questions and merge them nicely**

### Let's ask some questions

In [ ]:
chain('Do you provide job assistance and also do you provide job gurantee?')

In [17]:
chain("Do you guys provide internship and also do you offer EMI payments?")

{'query': 'Do you guys provide internship and also do you offer EMI payments?',
 'result': "Yes, we provide virtual internship and we don't offer EMI payments.",
 'source_documents': [Document(page_content='prompt: Do you provide any virtual internship?\nresponse: Yes', metadata={'source': 'Do you provide any virtual internship?', 'row': 14}),
  Document(page_content='prompt: Do we have an EMI option?\nresponse: No', metadata={'source': 'Do we have an EMI option?', 'row': 13}),
  Document(page_content='prompt: Do you provide any job assistance?\nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.', metadata={'source': 'Do you provide any job assistance?', 'row': 11}),
  Document(page_content='prompt: How can I contact the instructors for any doubts/support?\nresponse: We have created every lecture with a motive to explain everything in an easy-

In [18]:
chain("do you have javascript course?")

{'query': 'do you have javascript course?',
 'result': "I don't know.",
 'source_documents': [Document(page_content='prompt: I have never done programming and belong to a non-technical background. Can I take this course?\nresponse: Yes, this is the perfect course for anyone who has never done coding and wants to build a career in the IT/Data Analytics industry or just wants to perform better in their current job or business using data.', metadata={'source': 'I have never done programming and belong to a non-technical background. Can I take this course?', 'row': 24}),
  Document(page_content='prompt: I have never done programming in my life. Can I take this bootcamp?\nresponse: Yes, this is the perfect bootcamp for anyone who has never done coding and wants to build a career in the IT/Data Analytics industry or just wants to perform better in your current job or business using data.', metadata={'source': 'I have never done programming in my life. Can I take this bootcamp?', 'row': 0}),


In [19]:
chain("Do you have plans to launch blockchain course in future?")

{'query': 'Do you have plans to launch blockchain course in future?',
 'result': "I don't know.",
 'source_documents': [Document(page_content='prompt: Will the course be upgraded when there are new features in Power BI?\nresponse: Yes, the course will be upgraded periodically based on the new features in Power BI, and learners who have already bought this course will have free access to the upgrades.', metadata={'source': 'Will the course be upgraded when there are new features in Power BI?', 'row': 27}),
  Document(page_content='prompt: What business concepts and domains are covered in this course?\nresponse: We have covered the core functions such as Sales, Marketing, Finance, and Supply Chain with their fundamentals related to this course. The domain you will learn in this course is consumer goods which is projected to have more openings and high data analytics requirements at least until 2030.', metadata={'source': 'What business concepts and domains are covered in this course?', '

In [20]:
chain("should I learn power bi or tableau?")

{'query': 'should I learn power bi or tableau?',
 'result': 'This is a contextual question. If you are talking about a pure visualization tool Tableau is slightly better. Data connectors, modeling and transformation features are available in both. However, factually speaking Power BI is cheaper and offers tighter integration with the Microsoft environment. Since most companies use excel & Microsoft tools they start with Power BI or move towards Power BI for seamless integration with other Microsoft tools (called as Power platform). This makes the job openings grow at a much higher rate on Power BI and Power Platform. Also, Power BI has been leading the Gartner’s magic quadrant in BI for the last few years as the industry leader.',
 'source_documents': [Document(page_content='prompt: Power BI or Tableau which one is better?\nresponse: This is a contextual question. If you are talking about a pure visualization tool Tableau is slightly better. Data connectors, modeling and transformation

In [21]:
chain("I've a MAC computer. Can I use powerbi on it?")

{'query': "I've a MAC computer. Can I use powerbi on it?",
 'result': 'response: Hi\n\nPower BI desktop works only in Windows OS. Please look into the system requirements section on this page. However, you can use a virtual machine to install and work with Power BI in other Operating systems.',
 'source_documents': [Document(page_content='prompt: How can I use PowerBI on my Mac system?\nresponse: Hi\n\nYou can use VirtualBox to create a virtual machine and install Windows on it. This will allow you to run Power BI and Excel on your Mac.\n\nIf you\'re not familiar with setting up a virtual machine, there are many resources available on YouTube that can guide you through the process. Simply search for "installing virtual machines" and you\'ll find plenty of helpful videos.\n\nBest of luck with your studies!', metadata={'source': 'How can I use PowerBI on my Mac system?', 'row': 44}),
  Document(page_content='prompt: Does Power BI work in Mac OS/Ubuntu?\nresponse: Power BI desktop works o

In [ ]:
chain("I don't see power pivot. how can I enable it?")

In [ ]:
chain("What is the price of your machine learning course?")